lib/saluki-core/src/cache/mod.rs

```diff
@@ -0,0 +1,902 @@
+// TODO: If we're trying to add a reclaimed entry, and that entry comes immediately before any available capacity, and
+// the entry is aligned... we should skip actually tracking it as a reclaimed entry and instead simply decrement `len`,
+// adding the entry back to the available capacity at the end of the data buffer.
+//
+// This avoids specific fragmentation where the reclaimed entry might be too small for a string, requiring us to spill
+// to the buffer, and then we waste space doing so when we could have potentially had a more optimal usage if the
+// available capacity was simply larger.
+
+#![allow(dead_code)]
+use std::{
+    alloc::Layout,
+    collections::BTreeSet,
+    hash::Hasher as _,
+    num::NonZeroUsize,
+    ops::Deref,
+    ptr::NonNull,
+    sync::{
+        atomic::{
+            AtomicUsize,
+            Ordering::{AcqRel, Acquire},
+        },
+        Arc, Mutex,
+    },
+};
+
+const HEADER_LEN: usize = std::mem::size_of::<EntryHeader>();
+const HEADER_ALIGN: usize = std::mem::align_of::<EntryHeader>();
+
+/// An interned string.
+#[derive(Debug, Eq, PartialEq)]
+pub struct InternedString {
+    state: Arc<StringState>,
+}
+
+impl Deref for InternedString {
+    type Target = str;
+
+    fn deref(&self) -> &str {
+        self.state.get_entry()
+    }
+}
+
+#[derive(Debug)]
+struct StringState {
+    interner: Arc<Mutex<InternerState>>,
+    header: NonNull<EntryHeader>,
+}
+
+impl StringState {
+    fn get_entry<'a>(&'a self) -> &'a str {
+        // NOTE: We're specifically upholding a safety variant here by tying the lifetime of the string reference to our
+        // own lifetime, as the lifetime of the string reference *cannot* exceed the lifetime of the interner state,
+        // which we keep alive by holding `Arc<Mutex<InternerState>>`.
+        get_entry_string::<'a>(self.header.as_ptr())
+    }
+}
+
+impl Drop for StringState {
+    fn drop(&mut self) {
+        // TODO: We need to decrement the reference count on the entry, and if it's at zero, we try to add a tombstone
+        // to the interner state for it. That might fail if another caller is trying to intern the same string before we
+        // can get the lock to the interner state, but that's OK: we only care about at least _trying_ to tombstone it.
+
+        // SAFETY: We know that `self.header` is well-aligned for `EntryHeader` and is initialized.
+        let header = unsafe { self.header.as_ref() };
+        if header.refs.fetch_sub(1, AcqRel) == 1 {
+            // We decremented the reference count to zero, so try to mark this entry for reclamation.
+            let mut interner = self.interner.lock().unwrap();
```

Comment:
I need to delete this file, because it lives in `lib/stringtheory` now, but there's a comment in `InternerState::mark_for_reclamation` that goes over this.

Essentially:
- we only ever increment the active reference count for an entry when interning in `InternerState` (concurrent access mediated by a mutex)
- we only ever decrement the active reference count in `StringState::drop`
- only the last active reference (`StringState`) to a given entry can decrement it to zero
- when decremented to zero, `InternerState::mark_for_reclamation` is called, which checks to see if the entry is active (reference count > 0)

Since we can't call `InternerState::mark_for_reclamation` except when holding exclusive access, by the time we're in `mark_for_reclamation`, either:
- the caller was the last active reference, and we'll be able to reclaim the entry
- in between the decrementing the active reference count and obtaining the lock to `InternerState`, another caller tried to intern the same string, found the entry, incremented the active reference count, and when the call from `StringState::drop` finally obtains exclusive access and calls `mark_for_reclamation`, it will observe that the entry is actually (still) active, and do nothing